lib/vagrant-service-manager/installer.rb

```diff
@@ -0,0 +1,106 @@
+module VagrantPlugins
+  module ServiceManager
+    BINARY_NAME = {
+      docker: 'docker', openshift: 'oc'
+    }
+
+    class Installer
+      LABEL = 'servicemanager.commands.install_cli.message'
+      VERSION_CMD = {
+        docker: "docker version --format '{{.Server.Version}}'",
+        openshift: "oc version | grep 'oc' | grep -oE '[0-9.]+'"
+      }
+      BINARY_REGEX = {
+        windows: { docker: /\/docker.exe$/, openshift: 'TODO' },
+        unix: { docker: /\/docker$/, openshift: 'TODO' }
+      }
+
+      def initialize(type, machine, env, options)
+        if options[:box_version] == 'cdk'
+          ui.info I18n.t('servicemanager.commands.install_cli.unsupported_box')
+          exit 126
+          # TODO: Remove this condition when CDK need to be supported
+        end
+
+        @binary_exists = true
+        @url = ''
+        @archive_file_path = ''
+        @type = type
+        @machine = machine
+        @ui = env.ui
+        @temp_bin_dir = "#{ServiceManager.temp_dir}/#{@type.to_s}"
+        options['--cli-version'] ||= PluginUtil.execute_once(@machine, @ui, VERSION_CMD[@type])
+        @binary_object = BinaryHandlerFactory.build(machine, env, { type: @type }.merge(options))
+        @binary_name = @binary_object.binary_name
+        @path = options['--path'] || @binary_object.binary_path
+      end
+
+      def install
+        unless PluginUtil.binary_downloaded?(@path)
+          @binary_exists = false
+          set_download_url
+          set_archive_path
+          ensure_binary_and_temp_directories
+          download_archive
+          prepare_binary
+        end
+        print_message
+      end
+
+      private
+
+      def ensure_binary_and_temp_directories
+        FileUtils.mkdir_p(bin_dir) unless File.directory?(bin_dir)
+        FileUtils.mkdir_p(@temp_bin_dir) unless File.directory?(@temp_bin_dir)
+      end
+
+      def download_archive
+        unless File.exists? @archive_file_path
+          Vagrant::Util::Downloader.new(@url, @archive_file_path).download!
+        end
+      rescue Vagrant::Errors::DownloaderError => e
+        @ui.error e.message
+        exit 126
+      end
+
+      def prepare_binary
+        tmp_binary_file_path = @archive_file_path
+
+        if ['.tgz', '.tar.gz', '.zip'].include? File.extname(@archive_file_path)
+          archive_dir_name = @archive_file_path.sub(/.tgz|.tar.gz|.zip/, '')
+          tmp_binary_file_path = "#{archive_dir_name}/#{@binary_name}"
+          ArchiveFactory.build(@archive_file_path, tmp_binary_file_path, @ui, file_regex).unpack
+        end
+
+        FileUtils.cp(tmp_binary_file_path, @path)
+        File.chmod(0755, @path) unless Vagrant::Util::Platform.windows?
+      rescue Vagrant::Errors::VagrantError => e
+        @ui.error e.message
+        exit 126
+      end
+
+      def bin_dir
+        File.dirname(@path)
+      end
+
+      def set_download_url
+        @url = @binary_object.download_url
```

Comment:
I think here might be a code smell. Why don't you let all the file mangling happening the handlers. The handler knows the url. Give him the location where to download and let it do everything. Create a base class for the handlers to share common functionality between the Docker and OpenShift handler. I think that would be a cleaner design.